Add ML-KEM768 KATs from BoringSSL

andrewkdinh opened the pull request:
Add KATs for ML-KEM768 under CCLA from https://boringssl.googlesource.com/boringssl/

**Please advise for how I should credit BoringSSL in the test files.**

These KATs test key generation, encapsulation, and decapsulation for the ML-KEM768 provider.

Relevant notes:
- Added functionality to the ML-KEM provider to export/import. These may not be fully implemented yet (see https://github.com/openssl/project/issues/1036)
- Exposed some more low-level ML-KEM API's to allow for deterministic encapsulation/key generation
- Actually run 'mlkem_internal_test' with `make test`
- Fixes openssl/project#1045

<!--
Thank you for your pull request. Please review these requirements:

Contributors guide: https://github.com/openssl/openssl/blob/master/CONTRIBUTING

Other than that, provide a description above this comment if there isn't one already

If this fixes a github issue, make sure to have a line saying 'Fixes #XXXX' (without quotes) in the commit message.
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [x] tests are added or updated

t8m committed:
speed.c: Check for 0 block size

Although this cannot really happen check for 0 block size
to avoid division by 0.

Fixes Coverity 1633936

Reviewed-by: Paul Dale <ppzgs1@gmail.com>
Reviewed-by: Dmitry Belyavskiy <beldmit@gmail.com>
Reviewed-by: Matt Caswell <matt@openssl.org>
Reviewed-by: Saša Nedvědický <sashan@openssl.org>
(Merged from https://github.com/openssl/openssl/pull/25822)

andrewkdinh commented:
Tagging everybody from the [initial PR](https://github.openssl.org/mbaentsch/openssl/pull/5)

@mattcaswell @baentsch @slontis @paulidale @t8m

baentsch committed:
ML-KEM768

Based on code from BoringSSL covered under Google CCLA

Original code at https://boringssl.googlesource.com/boringssl/+/HEAD/crypto/mlkem

VSCode automatic formatting (andrewd@openssl.org)

Just do some basic formatting to make diffs easier to read later: convert from 2 to 4 spaces, add newlines after function declarations, and move function open curly brace to new line (andrewd@openssl.org)

Move variable init to beginning of each function (andrewd@openssl.org)

replace CBB API

fixing up constants and parameter lists

replace BORINGSSL_keccac calls with EVP calls

added library symbols and low-level test case

switch boringssl constant time routines for OpenSSL ones

data type assertion and negative test added

moved mlkem.h to include/crypto

changed function naming to be in line with ossl convention

remove Google license terms based on CCLA

add constant_time_lt_32

convert asserts to ossl_asserts where possible

add bssl keccak, pubK recreation, formatting

add provider interface to utilize mlkem768 code enabling TLS1.3 use

revert to OpenSSL DigestXOF

use EVP_MD_xof() to determine digest finalisation (pauli@openssl.org)

change APIs to return error codes; reference new IANA number; move static asserts to one place

remove boringssl keccak for good

fix coding style and return value checks

ANSI C compatibility changes

remove static cache objects

all internal retval functions used leading to some new retval functions

Reviewed-by: Tomas Mraz <tomas@openssl.org>
Reviewed-by: Matt Caswell <matt@openssl.org>
(Merged from https://github.com/openssl/openssl/pull/25848)

baentsch commented:
Thanks for this work @andrewkdinh ! Could you please re-base so the diff is easier to read?

t8m commented:
> Thanks for this work @andrewkdinh ! Could you please re-base so the diff is easier to read?

It would make sense to move this to openssl/openssl after the initial PR is merged into the feature branch.

andrewkdinh committed:
Add ML-KEM768 KATs from BoringSSL

Add KATs for ML-KEM768 under CCLA from https://boringssl.googlesource.com/boringssl/

These KATs test key generation, encapsulation, and decapsulation for the ML-KEM768 provider.

Relevant notes:
- Added functionality to the ML-KEM provider to export/import. These may not be fully implemented yet (see openssl/private#698)
- Exposed some more low-level ML-KEM API's to allow for deterministic encapsulation/key generation
- Actually run 'mlkem_internal_test' with `make test`
- Fixes openssl/private#704

andrewkdinh commented:
@baentsch I've re-based this now

@t8m Ok, I think I'll need to create a separate PR in openssl/openssl once https://github.com/openssl/openssl/pull/25848 is merged

baentsch commented on providers/implementations/keymgmt/mlkem_kmgmt.c:
```diff
@@ -507,13 +537,33 @@ static void *mlkem_gen(void *vctx, OSSL_CALLBACK *osslcb, void *cbarg)
+            || ossl_mlkem768_marshal_private_key(mkey->encoded_privkey,
+                                                 &mkey->privkey)
+            || ossl_mlkem768_public_from_private(&mkey->pubkey,
+                                                 &mkey->privkey))
```
marshal_public_key missing here?

andrewkdinh replied:
Line 559 already writes the encoded public key to `mkey->encoded_pubkey`

```c
if (!ossl_mlkem768_generate_key(mkey->encoded_pubkey, gctx->seed,
```

baentsch commented on providers/implementations/keymgmt/mlkem_kmgmt.c:
```diff
@@ -580,11 +634,12 @@ const OSSL_DISPATCH ossl_mlkem768_keymgmt_functions[] = {
-     *  { OSSL_FUNC_KEYMGMT_EXPORT, (void (*)(void))mlkem_import },
-     */
+     * TODO(ML-KEM): https://github.com/openssl/private/issues/698
+     * Export/import functionality has been partially implemented.
```
Worthwhile stating here which things are missing?

andrewkdinh replied:
The encode/decode logic pretty much matches what I saw in other providers. But from https://github.com/openssl/private/issues/698:

> prove interoperability -- and for that we need reference to a standard (please fill in the above if you can: I didn't find anything, so if we want to do this without a standard (?), we need to agree on/how to do interop testing with other implementations).

I haven't done anything like this

baentsch replied:
ACK. Suggest referencing to https://github.com/openssl/openssl/issues/25885 now (contains draft spec references).

baentsch commented on test/evp_test.c:
```diff
@@ -2288,30 +2346,75 @@ static int decapsulate(EVP_TEST *t, EVP_PKEY_CTX *ctx, const char *op,
-                          secret, secretlen);
+
+        /* Also attempt to decrypt if we have the private key */
+        if (found_key || kdata->encoded_priv_key != NULL)
```
Wouldn't it be better for the test to fail if there's no private key available?

andrewkdinh replied:
The tests for ML-KEM-768 encap only have the encoded public key (and no private key)

baentsch commented on test/recipes/30-test_evp.t:
```diff
@@ -130,6 +130,9 @@ my @defltfiles = qw(
                      evppkey_kdf_scrypt.txt
                      evppkey_kdf_tls1_prf.txt
                      evppkey_rsa.txt
+                     evppkey_mlkem768_keygen.txt
```
shouldn't they be done (pushed to defltfiles) only if no_mlkem is not set?

andrewkdinh replied:
Updated: https://github.com/baentsch/openssl/compare/b8e8d90a0a369a69f399885419df3d793006fe3a..3cc382434d4b8f576fc6213e5741b55ffe6ce396

baentsch replied:
Not quite OK: The new variable name is "no_mlkem" (not declared as such). After fixing that, the tests fail for me (both, test_evp and test_evp_extra): Do I need to do some special setup to run them? I used `make TESTS='test_evp_extra test_evp' test`...

andrewkdinh replied:
Oops, there was a couple incorrect things after fixing a rebase. All tests succeed for me when I run `make test`... can you try again with the latest commit?

andrewkdinh committed:
Don't run ML-KEM-768 tests if ML-KEM is disabled

baentsch commented on providers/implementations/keymgmt/mlkem_kmgmt.c:
```diff
@@ -507,13 +537,33 @@ static void *mlkem_gen(void *vctx, OSSL_CALLBACK *osslcb, void *cbarg)
+        debug_print("MLKEMKM generate random keys");
+        if (!ossl_mlkem768_generate_key(mkey->encoded_pubkey, gctx->seed,
+                                        &mkey->privkey, mkey->mlkem_ctx)
+            || ossl_mlkem768_marshal_private_key(mkey->encoded_privkey,
```
"!" missing -- wrong failure condition and the reason for "evp_extra" test failiing.

baentsch commented on providers/implementations/keymgmt/mlkem_kmgmt.c:
```diff
@@ -507,13 +537,33 @@ static void *mlkem_gen(void *vctx, OSSL_CALLBACK *osslcb, void *cbarg)
+                                        &mkey->privkey, mkey->mlkem_ctx)
+            || ossl_mlkem768_marshal_private_key(mkey->encoded_privkey,
+                                                 &mkey->privkey)
+            || ossl_mlkem768_public_from_private(&mkey->pubkey,
```
"!"

baentsch commented:
@andrewkdinh fyi, with the three code changes as per the above, the tests all work -- for ML-KEM. Failures in SM2 and FFDHE...

andrewkdinh commented:
@baentsch All tests pass for me with `make test`. But this can also be verified when I open the PR to openssl/openssl when CI runs

baentsch commented:
> @baentsch All tests pass for me with `make test`. But this can also be verified when I open the PR to openssl/openssl when CI runs

@andrewkdinh : Confirmed for ML-KEM. Thanks! I'd be all for opening this as a public PR now. One heads-up (nit, I hope), though: These 2 tests (unrelated to MLKEM) fail for me (just did clean build using `./config && make -j && make test` on a vanilla x64 ubuntu box):

```

        # INFO:  @ test/testutil/stanza.c:21
        # Reading ../../test/recipes/30-test_evp_data/evppkey_ffdhe.txt
        # INFO:  @ test/testutil/stanza.c:122
        # Starting "RFC7919 DH tests" tests at line 15
../../util/wrap.pl ../../test/evp_test -config ../../test/default-and-legacy.cnf ../../test/recipes/30-test_evp_data/evppkey_ffdhe.txt => 139
not ok 25 - running evp_test -config ../../test/default-and-legacy.cnf evppkey_ffdhe.txt
# ------------------------------------------------------------------------------
        # INFO:  @ test/testutil/stanza.c:21
        # Reading ../../test/recipes/30-test_evp_data/evppkey_sm2.txt
        # INFO:  @ test/testutil/stanza.c:122
        # Starting "SM2 tests" tests at line 19
        # INFO:  @ test/testutil/stanza.c:122
        # Starting "SM2 key generation tests" tests at line 78
../../util/wrap.pl ../../test/evp_test -config ../../test/default-and-legacy.cnf ../../test/recipes/30-test_evp_data/evppkey_sm2.txt => 139
not ok 78 - running evp_test -config ../../test/default-and-legacy.cnf evppkey_sm2.txt
```


@t8m @mattcaswell: What does it take to merge https://github.com/openssl/openssl/pull/25848 such as to allow @andrewkdinh to raise this code as a public PR towards the feature branch?

baentsch commented:
@andrewkdinh Please redirect this PR towards the mlkem feature branch now: I'm anxious to (first review, of course, and) land this code (and APIs and provider improvements) there to base the next steps on that (ml-kem-1024, probably to ascertain our approach "generalizes" well). Thanks in advance!

andrewkdinh commented:
@baentsch I need to wait until your PR is fully synced into feature/ml-kem so that the diff is just my changes (usually <24 hours): https://github.com/openssl/openssl/compare/feature/ml-kem...andrewkdinh:openssl:mlkem-kats

I'll also be looking into the failing tests; strange that it didn't fail for me locally

mattcaswell commented:
> @baentsch I need to wait until your PR is fully synced into feature/ml-kem so that the diff is just my changes (usually <24 hours): https://github.com/openssl/openssl/compare/feature/ml-kem...andrewkdinh:openssl:mlkem-kats

@baentsch's PR has been merged to the feature/ml-kem branch. You should be able to create your PR now.

andrewkdinh commented:
Opened https://github.com/openssl/openssl/pull/25938 to replace this PR
